src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,234 @@
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.has_orthogonal_rows` means matrix `A` has orthogonal
+    (with respect to `dot_product`) rows. -/
+def has_orthogonal_rows [fintype n] [has_mul α] [add_comm_monoid α]
+(A : matrix m n α) : Prop :=
+∀ ⦃i₁ i₂⦄, i₁ ≠ i₂ → dot_product (A i₁) (A i₂) = 0
+
+/-- `A.has_orthogonal_cols` means matrix `A` has orthogonal
+    (with respect to `dot_product`) columns. -/
+def has_orthogonal_cols [fintype m] [has_mul α] [add_comm_monoid α]
+(A : matrix m n α) : Prop :=
+has_orthogonal_rows Aᵀ
+
+/-- `Aᵀ` has orthogonal rows iff `A` has orthogonal columns. -/
+lemma transpose_has_orthogonal_rows_iff_has_orthogonal_cols
+[fintype m] [has_mul α] [add_comm_monoid α] (A : matrix m n α) :
+  Aᵀ.has_orthogonal_rows ↔ A.has_orthogonal_cols :=
+iff.rfl
+
+/-- `Aᵀ` has orthogonal columns iff `A` has orthogonal rows. -/
+lemma transpose_has_orthogonal_cols_iff_has_orthogonal_rows
+[fintype n] [has_mul α] [add_comm_monoid α] (A : matrix m n α) :
+  Aᵀ.has_orthogonal_cols ↔ A.has_orthogonal_rows :=
+by simp [has_orthogonal_rows, has_orthogonal_cols, transpose_row_eq_col]
+
+-----------------------------------------------------------------------------------
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d):=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h]}
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by { intros i j h, have h' := subsingleton.elim i j, contradiction }
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, by simp
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+by { intros i j h, simp [h] }
+
+@[simp] lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0):
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+@[simp] lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.add
+[add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
```

Comment:
Every line from `lemma` up to and including the `:=` should be indented. This applies throughout